websites/management/commands/markdown_cleaning/link_parser.py

```diff
@@ -0,0 +1,148 @@
+import json
+from dataclasses import dataclass, field
+
+from pyparsing import (
+    CharsNotIn,
+    Literal,
+    Optional,
+    ParseResults,
+    StringEnd,
+    White,
+    nestedExpr,
+    originalTextFor,
+    quotedString,
+)
+
+from websites.management.commands.markdown_cleaning.parsing_utils import (
+    WrappedParser,
+    unescape_quoted_string,
+)
+
+
+@dataclass
+class MarkdownLink:
+    """
+    Representation of a markdown link OR image.
+    """
+
+    text: str
+    destination: str
+    is_image: bool = False
+    title: str = ""
+
+    # Tuple of tuples (ParseResults, start_index, end_index) where
+    # - ParseResult has named tokens
+    #       - link: MarkdownLinkOrImage
+    #       - original_text: str
+    # - start_index, end_index are the start/end of this link within self.text
+    text_links: list = field(default_factory=tuple)
+
+    def to_markdown(self):
+        """Generate markdown representation of this link/image."""
+        prefix = "!" if self.is_image else ""
+        title_suffix = " " + json.dumps(self.title) if self.title else ""
+        return f"{prefix}[{self.text}]({self.destination}{title_suffix})"
+
+
+class LinkParser(WrappedParser):
```

Comment:
This is where the main new logic is. A brief look at https://spec.commonmark.org/0.30/#links might be worthwhile, too. This parser does not handle all the CommonMark edge cases, but should be good enough for us.

#### Aside: Why not use a python libraring build for Markdown parsing?
Most of the pyparsing stuff was introduced in #1137 but adding it here for some context...

A specific Markdown library might be useful. The two reasons I shied away from that for now are (A) shortcodes and (B) my impression is that with a Markdown parsing library, we would need to:
1. for a single page, parse the **whole** markdown into HTML
2. make changes on the HTML
3. convert back to markdown

That's a global transformation: Ie, we'd be changing converting and unconverting all of the markdown to make localized edits. In principal it should be fine... the conversion process *should* be lossless. But I suspect there are a lot of quirks involved... `-` vs `*` for lists, `_` vs `*` for italics, inclusion (or not) of extra whitespace... All of those things *shouldn't* matter, but... they might. And I'd rather not tempt fate.

In contrast, Pyparsing lets us make localized transformations of the markdown text. We can change a link in the middle and be confident that the pragraphs above and below did not change **at all**.